src/main/java/frc/robot/subsystems/swerve/Swerve.java

```diff
@@ -0,0 +1,286 @@
+		return commandsBuilder;
+	}
+
+	public SwerveConstants getConstants() {
+		return constants;
+	}
+
+	public SwerveStateHandler getStateHandler() {
+		return stateHandler;
+	}
+
+
+	public void configPathPlanner(Supplier<Pose2d> currentPoseSupplier, Consumer<Pose2d> resetPoseConsumer, RobotConfig robotConfig) {
+		PathPlannerUtils.configPathPlanner(
+			currentPoseSupplier,
+			resetPoseConsumer,
+			this::getRobotRelativeVelocity,
+			(speeds) -> driveByState(speeds, SwerveState.DEFAULT_PATH_PLANNER),
+			constants.pathPlannerHolonomicDriveController(),
+			robotConfig,
+			() -> !Field.isFieldConventionAlliance(),
+			this
+		);
+	}
+
+	public void setHeadingSupplier(Supplier<Rotation2d> headingSupplier) {
+		this.headingSupplier = headingSupplier;
+	}
+
+	public void setHeading(Rotation2d heading) {
+		gyro.setYaw(heading);
+		gyro.updateInputs(gyroSignals.yawSignal());
+		headingStabilizer.unlockTarget();
+		headingStabilizer.setTargetHeading(heading);
+	}
+
+	protected void resetPIDControllers() {
+		constants.xMetersPIDController().reset();
+		constants.yMetersPIDController().reset();
+		constants.rotationDegreesPIDController().reset();
+	}
+
+
+	public void update() {
+		gyro.updateInputs(gyroSignals.yawSignal());
+		modules.updateInputs();
+
+		currentState.log(constants.stateLogPath());
+
+		ChassisSpeeds fieldRelativeSpeeds = getFieldRelativeVelocity();
+		Logger.recordOutput(constants.velocityLogPath() + "Rotation", fieldRelativeSpeeds.omegaRadiansPerSecond);
+		Logger.recordOutput(constants.velocityLogPath() + "X", fieldRelativeSpeeds.vxMetersPerSecond);
+		Logger.recordOutput(constants.velocityLogPath() + "Y", fieldRelativeSpeeds.vyMetersPerSecond);
+		Logger.recordOutput(constants.velocityLogPath() + "Magnitude", SwerveMath.getDriveMagnitude(fieldRelativeSpeeds));
+
+		Logger.recordOutput(getLogPath() + "OdometrySamples", getNumberOfOdometrySamples());
+	}
+
+
+	public int getNumberOfOdometrySamples() {
+		return Math.min(gyroSignals.yawSignal().asArray().length, modules.getNumberOfOdometrySamples());
+	}
+
+	public OdometryObservation[] getAllOdometryObservations() {
+		OdometryObservation[] odometryObservations = new OdometryObservation[getNumberOfOdometrySamples()];
+
+		for (int i = 0; i < odometryObservations.length; i++) {
+			odometryObservations[i] = new OdometryObservation(
+				modules.getWheelPositions(i),
+				gyro instanceof EmptyGyro ? Optional.empty() : Optional.of(gyroSignals.yawSignal().asArray()[i]),
+				gyroSignals.yawSignal().getTimestamps()[i]
+			);
+		}
+
+		return odometryObservations;
+	}
+
+	public double getDriveRadiusMeters() {
+		Translation2d[] modulePositionsFromCenterMeters = modules.getModulePositionsFromCenterMeters();
+		double sum = 0;
+		for (Translation2d modulePositionFromCenterMeters : modulePositionsFromCenterMeters) {
+			sum += modulePositionFromCenterMeters.getDistance(new Translation2d());
+		}
+		return sum / modulePositionsFromCenterMeters.length;
+	}
```

Comment:
This whole thing should be a constant calculated at startup

- Change to constant calculated at startup
- change function to return constant / refactor code to use the constant